Meadowlark-js/packages/meadowlark-core/test/validation/EqualityConstraintValidator.test.ts

```diff
@@ -0,0 +1,248 @@
+ */
+function sectionDomainEntity(): DomainEntity {
+  const metaEd: MetaEdEnvironment = newMetaEdEnvironment();
+
+  MetaEdTextBuilder.build()
+    .withBeginNamespace('EdFi')
+    .withStartDomainEntity('Section')
+    .withDocumentation('doc')
+    .withStringIdentity('SectionIdentifier', 'doc', '30')
+    .withDomainEntityIdentity('CourseOffering', 'doc')
+    .withDomainEntityProperty('ClassPeriod', 'doc', true, true)
+    .withMergeDirective('ClassPeriod.School', 'CourseOffering.School')
+    .withEndDomainEntity()
+
+    .withStartDomainEntity('CourseOffering')
+    .withDocumentation('doc')
+    .withStringIdentity('LocalCourseCode', 'doc', '30')
+    .withDomainEntityIdentity('School', 'doc')
+    .withEndDomainEntity()
+
+    .withStartDomainEntity('ClassPeriod')
+    .withDocumentation('doc')
+    .withStringIdentity('ClassPeriodName', 'doc', '30')
+    .withDomainEntityIdentity('School', 'doc')
+    .withEndDomainEntity()
+
+    .withStartDomainEntity('School')
+    .withDocumentation('doc')
+    .withStringIdentity('SchoolId', 'doc', '30')
+    .withEndDomainEntity()
+    .withEndNamespace()
+    .sendToListener(new NamespaceBuilder(metaEd, []))
+    .sendToListener(new DomainEntityBuilder(metaEd, []));
+
+  domainEntityReferenceEnhancer(metaEd);
+  entityPropertyApiSchemaDataSetupEnhancer(metaEd);
+  entityApiSchemaDataSetupEnhancer(metaEd);
+  referenceComponentEnhancer(metaEd);
+  apiPropertyMappingEnhancer(metaEd);
+  propertyCollectingEnhancer(metaEd);
+  apiEntityMappingEnhancer(metaEd);
+  jsonPathsMappingEnhancer(metaEd);
+  equalityConstraintEnhancer(metaEd);
+
+  const namespace: Namespace = metaEd.namespace.get('EdFi') as Namespace;
+  return namespace.entity.domainEntity.get('Section') as DomainEntity;
+}
+
+describe('when CourseOffering schoolId is a mismatch with a single ClassPeriod', () => {
+  let validationResult: string[] = [];
+  const sectionBody = {
+    sectionIdentifier: 'sectionIdentifier',
+    courseOfferingReference: {
+      localCourseCode: 'localCourseCode',
+      schoolId: 6,
+      sessionName: 'sessionName',
+      schoolYear: 2020,
+    },
+    classPeriods: [
+      {
+        classPeriodReference: {
+          schoolId: 6666,
+          classPeriodName: 'classPeriodName1',
+        },
+      },
+    ],
+  };
+
+  beforeAll(() => {
+    const section: DomainEntity = sectionDomainEntity();
+    validationResult = validateEqualityConstraints(section, sectionBody);
+  });
+
+  it('should have a validation failure', () => {
+    expect(validationResult).toMatchInlineSnapshot(`
+      [
+        "Constraint failure: document paths $.classPeriods[*].classPeriodReference.schoolId and $.courseOfferingReference.schoolId must have the same values",
+      ]
+    `);
+  });
+});
+
+describe('when CourseOffering schoolId is a mismatch with two ClassPeriods', () => {
+  let validationResult: string[] = [];
+  const sectionBody = {
```

Comment:
Thinking about new developers, would this be a good location to add an explanatory note? Maybe something like "This document is invalid because the `schoolId` does not match between `courseOfferingReference` the other two instances. This is an example of usage of MetaEd's merge directive, whereby nested documents with the same keys must have matching values for those keys."